lib/byron/src/Cardano/Wallet/Byron/Network.hs

```diff
@@ -676,75 +698,114 @@ chainSyncWithBlocks
-                    respond (Just intersection)
-                    clientStIdle
-
-            , recvMsgIntersectNotFound = \_tip ->
-                ChainSyncClient $ do
-                    respond Nothing
-                    clientStIdle
+            -> P.ClientPipelinedStIntersect ByronBlock (Tip ByronBlock) m Void
+        clientStIntersect respond = P.ClientPipelinedStIntersect
+            { recvMsgIntersectFound = \intersection _tip -> do
+                respond (Just intersection)
+                clientStIdle oneByOne
+
+            , recvMsgIntersectNotFound = \_tip -> do
+                respond Nothing
+                clientStIdle oneByOne
             }
 
-        clientStNext
-            :: ([ByronBlock], Int)
-            -> (NextBlocksResult (Point ByronBlock) ByronBlock -> m ())
-            -> ClientStNext ByronBlock (Tip ByronBlock) m Void
-        clientStNext (blocks, n) respond
-            | n <= 1 = ClientStNext
-                { recvMsgRollBackward = onRollback
-                , recvMsgRollForward = \block tip ->
-                    ChainSyncClient $ do
-                        let cursor' = blockPoint block
-                        let blocks' = reverse (block:blocks)
-                        let tip'    = fromTip getGenesisBlockHash getEpochLength tip
-                        respond (RollForward cursor' tip' blocks')
-                        clientStIdle
-                }
-            | otherwise = ClientStNext
-                { recvMsgRollBackward = onRollback
-                , recvMsgRollForward = \block _ ->
-                    ChainSyncClient $ pure $ SendMsgRequestNext
-                        (clientStNext (block:blocks,n-1) respond)
-                        (pure $ clientStNext (block:blocks,1) respond)
-                }
-          where
-            onRollback point _tip = ChainSyncClient $ do
+        -- Simple strategy that sends a request and wait for the answer.
+        oneByOne
+            :: RequestNextStrategy m 'Z
+        oneByOne respond = P.SendMsgRequestNext
+            (collectResponses respond [] Zero)
+            (pure $ collectResponses respond [] Zero)
+
+        -- We only pipeline requests when it's safe to pipeline them. Safe
+        -- is when all blocks are considered 'stable' and immutable, i.e.
+        -- when we land more than `k` blocks from the tip, so when we start
+        -- more than `k + maxInFlight` blocks from the tip
+        --
+        --     0                       tip - k           tip
+        --     |---------------------------|-------------->
+        --               pipelined            one by one
+        --
+        pipeline
+            :: Nat n
+            -> RequestNextStrategy m n
+        pipeline (Succ n) respond | natToInt (Succ n) == fromIntegral maxInFlight =
+            P.CollectResponse Nothing $ collectResponses respond [] n
+        pipeline n respond =
+            P.SendMsgRequestNextPipelined $ pipeline (Succ n) respond
```

Comment:
I would use `CollectResponse (Just (SendMsgRequestNextPipelined ..)) ...`. This pokes the receiving queue and if we got something it will be available in the next step, so more promptly.  You could experiment with some strategies when to use `CollectResponse (Just ...) ..` and when `SendMsgRequestNextPipelined`.  But maybe you're not doing so because of `Jörmungandr` support?

Reply:
> But maybe you're not doing so because of Jörmungandr support?

Not really. We process blocks by batches through the wallet, to avoid having to trigger the database and create checkpoint after processing every block. The restoration loop is driving the networking side by asking for next blocks; it is in essence similar to how the ChainSync works; except that instead of request a single block, it requests a batch. 

So here, we simply leverage pipelining to collect the next batch more rapidly; it queues "maxInFlight" requests, and then collect them all. Before, we would simply call `RequestNext` several time, but waiting for each request to complete before calling the next one.

Reply:
I see; then you could pipline even more, collect and buffer things as they come from the network (on your side) and send batches for processing. If you do your own buffering you will be able to saturate the network.